topology/graph/linker.go

```diff
@@ -0,0 +1,219 @@
+	l.nodeEvent(node)
+}
+
+type DefaultLinker struct {
+}
+
+func (dl *DefaultLinker) GetABLinks(node *Node) []*Edge {
+	return nil
+}
+
+func (dl *DefaultLinker) GetBALinks(node *Node) []*Edge {
+	return nil
+}
+
+type ResourceLinker struct {
+	g          *Graph
+	abListener *listener
+	baListener *listener
+	glh1       GraphListenerHandler
+	glh2       GraphListenerHandler
+	linker     Linker
+	metadata   Metadata
+}
+
+func (rl *ResourceLinker) getLinks(node *Node, direction string) []*Edge {
+	metadata := Metadata{}
+	for k, v := range rl.metadata {
+		metadata[k] = v
+	}
+	metadata[direction] = string(node.ID)
+	return rl.g.GetNodeEdges(node, metadata)
+}
+
+func (rl *ResourceLinker) Start() {
+	if rl.glh1 != nil {
+		rl.abListener = &listener{
+			graph:        rl.g,
+			newLinksFunc: rl.linker.GetABLinks,
+			existingLinksFunc: func(node *Node) (edges []*Edge) {
+				return rl.getLinks(node, "Parent")
+			},
+			metadata: rl.metadata,
+		}
+		rl.glh1.AddEventListener(rl.abListener)
+	}
+
+	if rl.glh2 != nil {
+		rl.baListener = &listener{
+			graph:        rl.g,
+			newLinksFunc: rl.linker.GetBALinks,
+			existingLinksFunc: func(node *Node) (edges []*Edge) {
+				return rl.getLinks(node, "Child")
+			},
+			metadata: rl.metadata,
+		}
+		rl.glh2.AddEventListener(rl.baListener)
+	}
+}
+
+func (rl *ResourceLinker) Stop() {
+	if rl.glh1 != nil {
+		rl.glh1.RemoveEventListener(rl.abListener)
+	}
+
+	if rl.glh2 != nil {
+		rl.glh2.RemoveEventListener(rl.baListener)
+	}
+}
+
+func NewResourceLinker(g *Graph, glh1 GraphListenerHandler, glh2 GraphListenerHandler, linker Linker, m Metadata) *ResourceLinker {
+	rl := &ResourceLinker{
+		g:        g,
+		glh1:     glh1,
+		glh2:     glh2,
+		linker:   linker,
+		metadata: m,
+	}
+
+	return rl
+}
+
+// getFieldsAsArray returs an array of corresponding values from a field list
+func getFieldsAsArray(obj filters.Getter, fields []string) ([]interface{}, error) {
+	values := make([]interface{}, len(fields))
+	for i, index := range fields {
+		v, err := obj.GetField(index)
+		if err != nil {
+			return nil, err
+		}
+		values[i] = v
+	}
+	return values, nil
+}
+
+type MetadataIndexerLinker struct {
+	*ResourceLinker
+	indexer1     *MetadataIndexer
+	indexer2     *MetadataIndexer
+	edgeMetadata Metadata
+}
+
+func (mil *MetadataIndexerLinker) genID(parent, child *Node) Identifier {
+	args := []string{string(parent.ID), string(child.ID)}
+	for k, v := range mil.edgeMetadata {
+		args = append(args, k, v.(string))
+	}
+	return GenID(args...)
+}
+
+func (mil *MetadataIndexerLinker) GetABLinks(node *Node) (edges []*Edge) {
+	if fields, err := getFieldsAsArray(node, mil.indexer1.indexes); err == nil {
+		nodes, _ := mil.indexer2.Get(fields...)
+		for _, n := range nodes {
+			edges = append(edges, mil.g.CreateEdge(mil.genID(node, n), node, n, mil.edgeMetadata, time.Now(), ""))
+		}
+	}
+	return
+}
+
+func (mil *MetadataIndexerLinker) GetBALinks(node *Node) (edges []*Edge) {
```

Comment:
this is a code duplication of `GetABLinks()` if you use a utility function you can eliminate the duplication:

```
func (mil *MetadataIndexerLinker) GetUniLinks(node *Node, indexer1, indexer2 Indexer) (edges []*Edge) {
        ...
}

func (mil *MetadataIndexerLinker) GetABLinks(node *Node) (edges []*Edge) {
        retrun mil.GetUniLinks(node, mil.indexer1, mil.indexer2)
}

func (mil *MetadataIndexerLinker) GetBALinks(node *Node) (edges []*Edge) {
        retrun mil.GetUniLinks(node, mil.indexer2, mil.indexer1)
}
```

Reply:
As the links are not in the same direction, it would not really reduce code duplication

Reply:
see the updated implementation above - it eliminated the duplication

Reply:
With your version, the edges are created in the wrong direction. In one case, node should be the source node and in the other case it shoulbe the target node

Reply:
ok